File: src/main/kotlin/leetcode/string/ReArrangementLogFile.java
package leetcode.string;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class ReArrangementLogFile {
    /**  5ms */
    private String[] recorderFileList(String[] logs) {
        List<String> letterList = new ArrayList<>();
        List<String> digitList = new ArrayList<>();

        for (String log : logs) {
            if (Character.isDigit(log.split(" ")[1].charAt(0))) {
                digitList.add(log);
            } else {
                letterList.add(log);
            }
        }

        letterList.sort((s1, s2) -> {
            String[] s1x = s1.split(" ", 2);
            String[] s2x = s2.split(" ", 2);
            System.out.println("s1x: " + Arrays.toString(s1x));
            System.out.println("s2x: " + Arrays.toString(s2x));
            //문자 로그 사전순 비교
            int compared = s1x[1].compareTo(s2x[1]);
            System.out.println("compared: " + compared);

            // 비교가 같을 경우 :  0
            // 비교 대상의 순서가 앞으로 와야 할 경우 : 1
            // 비교 대상 순서가 뒤에 머 물러야 할 경우 : -1

            // 문자 동일한 경우 식별자 비교
            if (compared == 0){
                return s1x[0].compareTo(s2x[0]);
            } else{
                return compared;
            }
        });

        letterList.addAll(digitList);

        return letterList.toArray(new String[0]);
    }
    public static void main(String[] args) {
        String[] logs = {
                "dig1 8 1 5 1",
                "let1 art can",
                "dig2 3 6",
                "let2 own kit dig",
                "let3 art zero"
        };
        // Method 1
        ReArrangementLogFile re = new ReArrangementLogFile();
        String[] answer = re.recorderFileList(logs);

    }
}

File: src/main/kotlin/leetcode/string/rearrangementLogFile.kt
package leetcode.string


private fun reorderLogFiles(logs: Array<String>): Array<String> {
    // 1. 로그의 가장 앞부분은 식별자로서, 순서에 영향을 끼지치 않는다.
    // 2. 문자로 구성된 로그가 숫자 로그보다 앞에 오며, 문자 로그는 사전순으로 한다.
    // 3. 문자가 동일한 경우에는 식별자순으로 한다.
    // 4. 숫자 로그는 입력 순서대로 한다.
    val letterList = mutableListOf<String>()
    val digitList = mutableListOf<String>()

    logs.forEach { log ->
        if(Character.isDigit(log.split(" ")[1][0])){
            digitList.add(log)
        } else {
            letterList.add(log)
        }
    }

    letterList.sortWith(Comparator { s1: String , s2: String ->
        val s1x = s1.split(" ", limit = 2)
        val s2x = s2.split(" ", limit = 2)

        val compared = s1x[1].compareTo(s2x[1])

        if(compared == 0) {
            s1x[0].compareTo(s2x[0])
        } else {
            compared
        }
    })

    letterList.addAll(digitList)
    return letterList.toTypedArray()
}
fun main(){
    val logs = arrayOf(
        "dig1 8 1 5 1",
        "let1 art can",
        "dig2 3 6",
        "let2 own kit dig",
        "let3 art zero"
    )
    reorderLogFiles(logs)
}